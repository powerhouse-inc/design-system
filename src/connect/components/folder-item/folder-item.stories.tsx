import { Meta, StoryObj } from '@storybook/react';
import { FolderItem } from './folder-item';

const meta: Meta<typeof FolderItem> = {
    title: 'Connect/Components/FolderItem',
    component: FolderItem,
    argTypes: {
        onClick: { action: 'onClick' },
        onOptionsClick: { action: 'onOptionsClick' },
        title: { control: 'text' },
        mode: { control: { type: 'select', options: ['read', 'write'] } },
        onDragStart: { action: 'onDragStart' },
        onDragEnd: { action: 'onDragEnd' },
        onDropEvent: { action: 'onDropEvent' },
        item: { control: { type: 'object' } },
    },
    decorators: [
        Story => (
            <div className="w-[500px] bg-white p-10">
                <Story />
            </div>
        ),
    ],
};

export default meta;

type Story = StoryObj<typeof meta>;

export const ReadMode: Story = {
    args: {
        title: 'Chronicle Labs Chronicle Labs Chronicle Labs Chronicle Labs Chronicle Labs Chronicle Labs',
        item: {
            id: '1',
            label: 'Test Folder',
            availableOffline: false,
            path: '',
            type: 'FOLDER',
            syncStatus: 'SYNCING',
        },
    },
};

export const WriteMode: Story = {
    ...ReadMode,
    args: {
        ...ReadMode.args,
        mode: 'write',
    },
};

export const NotAllowedToCreateDocuments: Story = {
    ...ReadMode,
    args: {
        ...ReadMode.args,
        isAllowedToCreateDocuments: false,
    },
};
