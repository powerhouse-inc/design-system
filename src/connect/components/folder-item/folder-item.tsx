import {
    ADD_INVALID_TRIGGER,
    ADD_TRIGGER,
    ConnectDropdownMenu,
    DELETE,
    DropdownMenuHandlers,
    DUPLICATE,
    FOLDER,
    NodeInput,
    NodeOption,
    nodeOptionsMap,
    NodeProps,
    READ,
    REMOVE_TRIGGER,
    RENAME,
    SyncStatusIcon,
    TUiNodesContext,
    UiFolderNode,
    useDrag,
    useDrop,
    WRITE,
} from '@/connect';
import { Icon } from '@/powerhouse';
import { useState } from 'react';
import { twMerge } from 'tailwind-merge';

export type FolderItemProps = TUiNodesContext &
    NodeProps & {
        uiNode: UiFolderNode;
        className?: string;
    };

export function FolderItem(props: FolderItemProps) {
    const {
        uiNode,
        isAllowedToCreateDocuments,
        nodeOptions,
        isRemoteDrive,
        className,
        setSelectedNode,
        onRenameNode,
        onDuplicateNode,
        onDeleteNode,
        onAddTrigger,
        onRemoveTrigger,
        onAddInvalidTrigger,
    } = props;
    const [mode, setMode] = useState<typeof READ | typeof WRITE>(READ);
    const [isDropdownMenuOpen, setIsDropdownMenuOpen] = useState(false);
    const { dragProps } = useDrag(props);
    const { isDropTarget, dropProps } = useDrop(props);

    const isReadMode = mode === READ;

    function onCancel() {
        setMode(READ);
    }

    function onSubmit(name: string) {
        onRenameNode(name, uiNode);
    }

    function onClick() {
        setSelectedNode(uiNode);
    }

    const dropdownMenuHandlers: DropdownMenuHandlers = {
        [DUPLICATE]: () => onDuplicateNode(uiNode),
        [RENAME]: () => setMode(WRITE),
        [DELETE]: () => onDeleteNode(uiNode),
        [ADD_TRIGGER]: () => onAddTrigger(uiNode.driveId),
        [REMOVE_TRIGGER]: () => onRemoveTrigger(uiNode.driveId),
        [ADD_INVALID_TRIGGER]: () => onAddInvalidTrigger(uiNode.driveId),
    } as const;

    const nodeOptionsForKind = nodeOptions[uiNode.sharingType][FOLDER];

    const dropdownMenuOptions = Object.entries(nodeOptionsMap)
        .map(([id, option]) => ({
            ...option,
            id: id as NodeOption,
        }))
        .filter(option => nodeOptionsForKind.includes(option.id));

    function onDropdownMenuOptionClick(itemId: NodeOption) {
        const handler = dropdownMenuHandlers[itemId];
        if (!handler) {
            console.error(`No handler found for dropdown menu item: ${itemId}`);
            return;
        }
        handler();
        setIsDropdownMenuOpen(false);
    }

    const content =
        isReadMode || !isAllowedToCreateDocuments ? (
            <div className="ml-3 max-h-6 truncate font-medium text-slate-200">
                {uiNode.name}
            </div>
        ) : (
            <NodeInput
                className="ml-3 font-medium"
                defaultValue={uiNode.name}
                onCancel={onCancel}
                onSubmit={onSubmit}
            />
        );

    const textStyles = isReadMode
        ? 'text-gray-600 hover:text-gray-800'
        : 'text-gray-800';

    const containerStyles = twMerge(
        'group flex h-12 cursor-pointer select-none items-center rounded-lg bg-gray-200 px-2',
        textStyles,
        className,
        isDropTarget && 'bg-blue-100',
    );

    return (
        <div onClick={onClick} className="relative w-64">
            <div {...dragProps} {...dropProps} className={containerStyles}>
                <div className="flex items-center overflow-hidden">
                    <div className="p-1">
                        <div className="relative">
                            <Icon name="folder-close" size={24} />
                            {isReadMode &&
                                isRemoteDrive &&
                                uiNode.syncStatus && (
                                    <div className="absolute bottom-[-3px] right-[-2px] size-3 rounded-full bg-white">
                                        <div className="absolute left-[-2px] top-[-2px]">
                                            <SyncStatusIcon
                                                syncStatus={uiNode.syncStatus}
                                                overrideSyncIcons={{
                                                    SUCCESS:
                                                        'check-circle-fill',
                                                }}
                                            />
                                        </div>
                                    </div>
                                )}
                        </div>
                    </div>
                    {content}
                </div>
                {isReadMode && isAllowedToCreateDocuments && (
                    <ConnectDropdownMenu
                        open={isDropdownMenuOpen}
                        onOpenChange={setIsDropdownMenuOpen}
                        onItemClick={onDropdownMenuOptionClick}
                        items={dropdownMenuOptions}
                    >
                        <button
                            onClick={e => {
                                e.stopPropagation();
                                setIsDropdownMenuOpen(true);
                            }}
                            className={twMerge(
                                'ml-auto hidden group-hover:block',
                                isDropdownMenuOpen && 'block',
                            )}
                        >
                            <Icon
                                name="vertical-dots"
                                className="text-gray-600"
                            />
                        </button>
                    </ConnectDropdownMenu>
                )}
            </div>
        </div>
    );
}
