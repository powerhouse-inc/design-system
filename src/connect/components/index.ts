export * from './drive-view';
export * from './dropdown-menu';
export * from './sidebar';
export * from './tree-view';
export * from './tree-view-item';
