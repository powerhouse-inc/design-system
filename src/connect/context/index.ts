export * from './ItemsContext';
export * from './WagmiContext';
