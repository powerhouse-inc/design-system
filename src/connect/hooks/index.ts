export * from './tree-view';
export * from './useENSInfo';
