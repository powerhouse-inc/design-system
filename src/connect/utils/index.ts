export * from './address';
export * from './get-folder-status';
export * from './mocks';
export * from './node-options';
