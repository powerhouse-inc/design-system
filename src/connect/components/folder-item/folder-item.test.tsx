import { TreeItem } from '@/connect';
import { fireEvent, render, screen } from '@testing-library/react';
import { FolderItem } from './folder-item';

const item: TreeItem = {
    id: '1',
    parentFolder: null,
    label: 'Test Folder',
    availableOffline: false,
    path: '',
    type: 'FOLDER',
};

describe('FolderItem Component', () => {
    it('should match read snapshot', () => {
        const { asFragment } = render(
            <FolderItem
                mode="read"
                title="Chronicle Labs"
                onClick={() => {}}
                onOptionsClick={() => {}}
                onCancelInput={() => {}}
                onSubmitInput={() => {}}
                item={item}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should match write snapshot', () => {
        const { asFragment } = render(
            <FolderItem
                mode="write"
                title="Chronicle Labs"
                onClick={() => {}}
                onOptionsClick={() => {}}
                onCancelInput={() => {}}
                onSubmitInput={() => {}}
                item={item}
            />,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it('should call onClick handler', () => {
        const onClick = vi.fn();

        render(
            <FolderItem
                title="Chronicle Labs"
                onClick={onClick}
                onOptionsClick={() => {}}
                onCancelInput={() => {}}
                onSubmitInput={() => {}}
                item={item}
            />,
        );

        fireEvent.click(screen.getByText('Chronicle Labs'));
        expect(onClick).toBeCalledTimes(1);
    });
});
