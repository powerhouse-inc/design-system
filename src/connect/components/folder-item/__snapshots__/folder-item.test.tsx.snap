// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FolderItem Component > should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="group flex h-12 cursor-pointer select-none items-center rounded-lg bg-gray-200 px-2 text-gray-600 hover:text-gray-800"
  >
    <div
      class="relative flex flex-1 flex-row items-center overflow-hidden"
    >
      <div
        class="p-1"
      >
        <svg
          style="width: 24px; height: 24px;"
        >
          <use
            href="icons.svg#folder-close"
          />
        </svg>
      </div>
      <div
        class="ml-3 max-h-6 overflow-hidden whitespace-nowrap font-medium text-slate-200 group-hover:text-gray-800"
      >
        Chronicle Labs
      </div>
      <div
        class="absolute right-0 h-full w-12 bg-gradient-to-r from-transparent to-gray-200"
      />
    </div>
    <div>
      <svg
        class="hidden group-hover:inline-block"
        style="width: 24px; height: 24px;"
      >
        <use
          href="icons.svg#vertical-dots"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
